cmd/docker/docker.go

```diff
@@ -352,7 +353,10 @@ func runDocker(ctx context.Context, dockerCli *command.DockerCli) error {
 	// We've parsed global args already, so reset args to those
 	// which remain.
 	cmd.SetArgs(args)
-	err = cmd.Execute()
+	err = cmd.ExecuteContext(ctx)
+	if err != nil {
+		return err
```

Comment:
Did this get borked during a rebase? It's returning early, should be returning after hooks get called I think

Reply:
Oh! Should it still handle context errors early though? (cancelled context? or timeout?) I guess in that case the hooks should _not_ be called?

Reply:
I think it's fine to do it regardless in any case here? The plugin can decide if it wants to print something or not. Even if it's cancelled, a hook might want to add some hint/explain something.

Reply:
Yeah, not sure if a cancelled context should still be used; isn't context cancelation basically; "cancel my request" (on the first possible occasion) ? In some cases that may mean some work is needed to complete the transaction (flush to disk, e.g.) but calling hooks (and execute other binaries) feels like that should not be part of that.

Reply:
that's valid, sure. @Benehiko wdyt

Reply:
mmh i'm not sure if hooks should be called after a cancellation, it'd feel wrong to me

Reply:
I don't have too much opinion on this either way. We currently execute the hooks even when the cobra `Execute` function returns an error (which is a similar thing to context cancellation).

Cancelling the context early will still execute the cobra commands, but the commands checking for context cancellation will just return (with maybe an error?). 

See below an example of a cobra command not acting on the context cancellation. Even if it was acting and returning an error it will still execute.


```go
package main

import (
	"context"
	"fmt"

	"github.com/spf13/cobra"
)

func main() {
	ctx, cancel := context.WithCancel(context.Background())

	rootCmd := &cobra.Command{
		Use: "",
		RunE: func(cmd *cobra.Command, args []string) error {
			cmd.Help()
			fmt.Printf("Cobra command ctx: %v\n", cmd.Context().Err())
			return nil
		},
	}

	cancel()

	fmt.Printf("Before Cobra ctx: %v\n", ctx.Err())

	err := rootCmd.ExecuteContext(ctx)
	fmt.Printf("%v\n", err)
}
```

```sh
⋊> ~/G/test-cobra go run .                        09:14:06
Before Cobra ctx: context canceled
Usage:
   [flags]

Flags:
  -h, --help   help for this command
Cobra command ctx: context canceled
<nil>
```

Reply:
See here what the code currently looks like on the master branch
https://github.com/docker/cli/blob/master/cmd/docker/docker.go#L355-L367